requests-for-comments/2023-11-22-scheduled-flag-changes.md

```diff
@@ -0,0 +1,84 @@
+# Request for comments: Scheduled Feature Flag changes
+
+## Problem
+
+Currently, users can make updates to feature flags only manually and in real-time. This is not ideal, as it potentially leads to overlooked or forgotten changes, particularly when they fall outside normal working hours. This can lead to poorly timed releases, negatively impacting end users' engagement and feedback.
+
+## Solution
+
+Scheduling feature flag changes allows users to automatically change flag properties at a future point in time. For example, a user can schedule an initial release of the flag for internal testing, then enable it to the beta users group 2 days later, and finally roll it out to 100% of the users in 2 weeks.
+
+## Scope
+
+This feature should allow the following scheduled modifications to a feature flag:
+
+-   **Enable/disable the flag** (field: `is_active`)
+-   **Add rollout conditions to the flag** (field: `filters`)
+
+We want to be able to _generalize_ this implementation to other models such as `Survey` or `Experiment`.
+
+## Implementation
+
+To schedule future changes, we can have a scheduling API, specifying the model the operate on, the id, the changes to be executed, and the schedule time.
+
+```
+POST /api/projects/:project_id/schedule
+{
+    "model": "FeatureFlag",
+    "id": 57,
+    "changes": [ ... ]
+    "scheduled_at": <timestamp>
+}
+```
+
+A change to the record can be expressed as a JSON, specifying the field to be modified and the new value to be set:
+
+**Example 1: Enable the flag**
+
+```
+{
```

Comment:
I think it's worth keeping this more generalisable: let the models determine what to do with the payload, and the scheduler just blindly passes the json.

So, I'd propose a top-level in this json to be two things: "operation" & "payload", and inside "payload" is what you have here. "operation" could be "database_update" for example.

This makes it much easier to extend to other scheduled things that aren't necessarily represented by a model update. And if we don't extend it at all, it's still fine, we're not adding massive generalising complexity here.

Reply:
Right, so the method for updating the record will live in the model. The Celery task will pick the appropriate model and pass the JSON to the model's method. Correct?

Curious, what's an example of scheduling where we're not performing a database update?

Reply:
>  The Celery task will pick the appropriate model and pass the JSON to the model's method. Correct?

Yep, one slight change: pass it to a function on the model, let's say called `scheduled_changes_dispatcher()`, and every model that needs scheduled tasks implements this function. And then the scheduler doesn't care what happens inside, be it a model update, or something else.

I think most things would do _some_ database update, but it might not be the only thing. For random example: a scheduled cohort computation: it may or may not update the database, but all it needs to do is recalculate the cohort. A bit contrived example, but hopefully gets the point across. It becomes much harder to do side-effects for example, if the only thing the scheduled task can do is perform db updates.

Reply:
Makes sense. Another example of a side effect might be calling a 3rd party API?

One more thing - can I expect every scheduling scenario to be related to _some_ model?

Reply:
yes that makes sense

Reply:
One more thing regarding updating the record and the queue item. To reiterate, we want to _atomically_:
- Update the feature flag
- Mark the queue item as done (by setting the `executed_at` timestamp)

However, if the responsibility of updating the record is now on the `scheduled_changes_dispatcher()` method, this method must handle the transaction. Does it make sense that the queue item's update be managed by the model rather than the Celery task?

Reply:
> Does it make sense that the queue item's update be managed by the model rather than the Celery task?

No this is no good! All scheduling logic & scheduler table updates should be in the task itself. We lose the general isolation if `scheduled_changes_dispatcher` implementation needs to handle this.

> the scheduled_changes_dispatcher() method, this method must handle the transaction.
This isn't necessary I think? The celery task can start a transaction, and whatever happens inside `scheduled_changes_dispatcher()` is a part of the transaction, without it handling the transaction itself. For an example, see this: https://github.com/PostHog/posthog/blob/master/posthog/models/feature_flag/flag_matching.py#L726 - it's a context manager that starts a transaction, does some updates, without caring what happens inside.

Given we don't really support long running tasks here, I think it's okay to have this all in one transaction. It's a bit iffy wr.t. side effects, but I think an ok trade off to make, given the kind of things we want to schedule here are limited & aren't building a general scheduler for everything.

Reply:
great, makes sense, exactly what I'm looking for!